File: sidecar/raylab_sidecar/hardware.py
from __future__ import annotations

import json
import platform
import shutil
import subprocess

import psutil

from .models import HardwareInfo


def detect_hardware() -> HardwareInfo:
    gpu_names, gpu_type, source = _detect_gpus()
    memory_total_gb = _system_memory_gb()
    gpu_memory_total_gb = _gpu_memory_gb(gpu_type, memory_total_gb)
    return HardwareInfo(
        cpu_logical=psutil.cpu_count(logical=True) or 0,
        cpu_physical=psutil.cpu_count(logical=False),
        memory_total_gb=memory_total_gb,
        gpu_count=len(gpu_names),
        gpu_names=gpu_names,
        gpu_type=gpu_type,
        gpu_memory_total_gb=gpu_memory_total_gb,
        gpu_memory_shared=gpu_type == "mps" and gpu_memory_total_gb is not None,
        source=source,
    )


def _system_memory_gb() -> float:
    return round(psutil.virtual_memory().total / 1024 / 1024 / 1024, 1)


def _gpu_memory_gb(gpu_type: str, system_memory_gb: float) -> float | None:
    if gpu_type == "cuda":
        return _nvidia_gpu_memory_gb()
    if gpu_type == "mps":
        return system_memory_gb
    return None


def _detect_gpus() -> tuple[list[str], str, str]:
    nvidia = _nvidia_gpus()
    if nvidia:
        return nvidia, "cuda", "nvidia-smi"

    mps = _mac_mps_gpus()
    if mps:
        return mps, "mps", "system_profiler"

    ray = _ray_gpus()
    if ray:
        return ray, "ray", "ray"

    return [], "none", "psutil"


def _nvidia_gpus() -> list[str]:
    if not shutil.which("nvidia-smi"):
        return []
    try:
        result = subprocess.run(
            ["nvidia-smi", "--query-gpu=name", "--format=csv,noheader"],
            check=False,
            capture_output=True,
            text=True,
            timeout=5,
        )
    except (OSError, subprocess.TimeoutExpired):
        return []
    if result.returncode != 0:
        return []
    return [line.strip() for line in result.stdout.splitlines() if line.strip()]


def _nvidia_gpu_memory_gb() -> float | None:
    if not shutil.which("nvidia-smi"):
        return None
    try:
        result = subprocess.run(
            ["nvidia-smi", "--query-gpu=memory.total", "--format=csv,noheader,nounits"],
            check=False,
            capture_output=True,
            text=True,
            timeout=5,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    if result.returncode != 0:
        return None

    total_mib = 0.0
    for line in result.stdout.splitlines():
        try:
            total_mib += float(line.strip())
        except ValueError:
            continue
    if total_mib <= 0:
        return None
    return round(total_mib / 1024, 1)


def _mac_mps_gpus() -> list[str]:
    if platform.system() != "Darwin":
        return []
    if not shutil.which("system_profiler"):
        return []
    try:
        result = subprocess.run(
            ["system_profiler", "SPDisplaysDataType", "-json"],
            check=False,
            capture_output=True,
            text=True,
            timeout=10,
        )
    except subprocess.TimeoutExpired:
        return []
    if result.returncode != 0:
        return []
    try:
        payload = json.loads(result.stdout.strip() or "{}")
    except json.JSONDecodeError:
        return []

    displays = payload.get("SPDisplaysDataType", [])
    if not isinstance(displays, list):
        return []

    names: list[str] = []
    for item in displays:
        if not isinstance(item, dict):
            continue
        metal = str(item.get("spdisplays_metal") or item.get("sppci_metal") or "").lower()
        chipset = str(item.get("sppci_model") or item.get("spdisplays_chipset") or item.get("_name") or "").strip()
        if "supported" in metal or chipset.lower().startswith("apple "):
            names.append(f"Apple Metal/MPS ({chipset or 'GPU'})")
    return names


def _ray_gpus() -> list[str]:
    try:
        script = "import ray._private.utils as u, json; print(json.dumps(u.get_visible_accelerator_ids().get('GPU', [])))"
        result = subprocess.run(["python3", "-c", script], check=False, capture_output=True, text=True, timeout=5)
    except (OSError, subprocess.TimeoutExpired):
        return []
    if result.returncode != 0:
        return []
    try:
        ids = json.loads(result.stdout.strip() or "[]")
    except json.JSONDecodeError:
        return []
    if not ids:
        return []
    return [f"GPU {item}" for item in ids]

File: electron/src/hardware.js
'use strict';

const os = require('os');
const { spawnSync } = require('child_process');
const { which } = require('./bootstrap');

// ─── Hardware detection ───────────────────────────────────────────────────────

async function detectHardware() {
  const cpuLogical = os.cpus().length;
  const memoryTotalGb = parseFloat((os.totalmem() / (1024 ** 3)).toFixed(1));

  let cpuPhysical = null;
  try {
    const si = require('systeminformation');
    const cpuInfo = await si.cpu();
    cpuPhysical = cpuInfo.physicalCores || null;
  } catch (_) {}

  const { gpuNames, gpuType, source } = _detectGpus();
  const gpuMemoryTotalGb = _gpuMemoryGb(gpuType, memoryTotalGb);

  return {
    cpu_logical: cpuLogical,
    cpu_physical: cpuPhysical,
    memory_total_gb: memoryTotalGb,
    gpu_count: gpuNames.length,
    gpu_names: gpuNames,
    gpu_type: gpuType,
    gpu_memory_total_gb: gpuMemoryTotalGb,
    gpu_memory_shared: gpuType === 'mps' && gpuMemoryTotalGb !== null,
    source,
  };
}

function _detectGpus() {
  // 1. NVIDIA
  const nvidiaSmi = which('nvidia-smi');
  if (nvidiaSmi) {
    try {
      const result = spawnSync(nvidiaSmi, ['--query-gpu=name', '--format=csv,noheader'], {
        timeout: 5000, encoding: 'utf8',
      });
      if (result.status === 0 && result.stdout.trim()) {
        const names = result.stdout.trim().split('\n').map((s) => s.trim()).filter(Boolean);
        if (names.length > 0) return { gpuNames: names, gpuType: 'cuda', source: 'nvidia-smi' };
      }
    } catch (_) {}
  }

  // 2. Apple MPS
  if (process.platform === 'darwin') {
    const profiler = '/usr/sbin/system_profiler';
    try {
      const result = spawnSync(profiler, ['SPDisplaysDataType', '-json'], {
        timeout: 10000, encoding: 'utf8',
      });
      if (result.status === 0) {
        const data = JSON.parse(result.stdout);
        const displays = data.SPDisplaysDataType || [];
        const gpuNames = [];
        for (const item of displays) {
          const hasMetal =
            (item.spdisplays_metal || '').toLowerCase().includes('supported') ||
            (item.sppci_metal || '').toLowerCase().includes('supported');
          const isApple =
            (item.sppci_model || item.spdisplays_chipset || item._name || '')
              .toLowerCase().startsWith('apple ');
          if (hasMetal || isApple) {
            const label = item.sppci_model || item.spdisplays_chipset || item._name || 'GPU';
            gpuNames.push(`Apple Metal/MPS (${label})`);
          }
        }
        if (gpuNames.length > 0) return { gpuNames, gpuType: 'mps', source: 'system_profiler' };
      }
    } catch (_) {}
  }

  // 3. Ray fallback
  const python3 = which('python3') || 'python3';
  try {
    const result = spawnSync(
      python3,
      ['-c', "import ray._private.utils as u, json; print(json.dumps(u.get_visible_accelerator_ids().get('GPU', [])))"],
      { timeout: 5000, encoding: 'utf8' }
    );
    if (result.status === 0) {
      const ids = JSON.parse(result.stdout.trim());
      if (Array.isArray(ids) && ids.length > 0) {
        return { gpuNames: ids.map((id) => `GPU ${id}`), gpuType: 'ray', source: 'ray' };
      }
    }
  } catch (_) {}

  return { gpuNames: [], gpuType: 'none', source: 'psutil' };
}

function _gpuMemoryGb(gpuType, systemMemoryGb) {
  if (gpuType === 'none') return null;

  if (gpuType === 'cuda') {
    const nvidiaSmi = which('nvidia-smi');
    if (nvidiaSmi) {
      try {
        const result = spawnSync(
          nvidiaSmi,
          ['--query-gpu=memory.total', '--format=csv,noheader,nounits'],
          { timeout: 5000, encoding: 'utf8' }
        );
        if (result.status === 0 && result.stdout.trim()) {
          const totalMiB = result.stdout.trim().split('\n')
            .map((l) => parseFloat(l.trim()))
            .filter((n) => !isNaN(n))
            .reduce((a, b) => a + b, 0);
          if (totalMiB > 0) return parseFloat((totalMiB / 1024).toFixed(1));
        }
      } catch (_) {}
    }
  }

  if (gpuType === 'mps') return systemMemoryGb;  // shared memory

  return null;
}

module.exports = { detectHardware };
